README
NAME

    XML::Encoding - A perl module for parsing XML encoding maps.

SYNOPSIS

    use XML::Encoding;
    my $em_parser = new XML::Encoding(ErrorContext  => 2,
                                      ExpatRequired => 1,
                                      PushPrefixFcn => \&push_prefix,
                                      PopPrefixFcn  => \&pop_prefix,
                                      RangeSetFcn   => \&range_set);

    my $encmap_name = $em_parser->parsefile($ARGV[0]);

DESCRIPTION

    This module, which is built as a subclass of XML::Parser, provides a parser
    for encoding map files, which are XML files. The file maps/encmap.dtd in the
    distribution describes the structure of these files. Calling a parse method
    returns the name of the encoding map (obtained from the name attribute of
    the root element). The contents of the map are processed through the
    callback functions push_prefix, pop_prefix, and range_set.

INSTALLATION

    See the INSTALL file.

COPYRIGHT

    Copyright (C) 1998 Clark Cooper. All rights reserved.
    Copyright (C) 2007-2010, 2014, 2020 Steve Hay. All rights reserved.

LICENCE

    This distribution is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, i.e. under the terms of either the GNU
    General Public License or the Artistic License, as specified in the LICENCE
    file.


Changes
Revision history for Perl distribution XML-Encoding

2.12 Development

    - TODO

2.11 2020-12-08

    - Fill in release dates in scripts.

2.10 2020-12-07

    - Added optional Changes testing (skipped unless AUTHOR_TESTING).

    - Reformatted Changes file as per CPAN::Changes::Spec.

    - Made Perl Critic, Pod and Pod Coverage tests skip unless AUTHOR_TESTING.

    - Added optional POD coverage testing.

    - Added optional POD testing.

    - Added optional Perl::Critic testing.

    - Made code Perl::Critic clean.

    - Simplified Makefile.PL requirements.

2.09 2014-05-30

    - Uploaded source code to GitHub and included repository URLs in metadata.

    - Included META.json file in addition to META.yml.

    - Set minimum required ExtUtils::MakeMaker version to 6.64 to ensure that
      all parameters used are supported, to save jumping through hoops to
      support earlier versions.  (This should not be a problem since
      ExtUtils::MakeMaker 6.64 is easily installed into Perl 5.8.1 and above,
      that being the whole point of the new choice of minimum supported Perl
      version.)

    - Set minimum required Perl version to 5.8.1.  This is in line with the
      minimum requirement of the "Perl Toolchain".

2.08 2010-09-20

    - Changed test script to use Test::More, renamed it to test.t and moved it
      into the t/ folder.  [CPAN RT#61369]

    - Moved the perl module into the lib/ folder.

2.07 2009-01-29

    - Minor correction to POD.  [Daniel Leidert <daniel.leidert@wgdd.de>]

2.06 2009-01-20

    - Fixed make_encmap to avoid spurious output when there is an error with
      the input file.  [CPAN RT#41854]

2.05 2008-09-11

    - Changed the shebang line in the two scripts to be less specific.  [Daniel
      Leidert <daniel.leidert@wgdd.de>]

2.04 2008-09-06

    - Updated the maps iso-8859-2.xml through iso-8859-9.xml and added the maps
      iso-8859-1.xml, iso-8859-10.xml, iso-8859-11.xml, iso-8859-13.xml through
      iso-8859-16.xml, and windows-1251.xml.  [Daniel Leidert
      <daniel.leidert@wgdd.de>]

    - Updated the map windows-1250.xml and added the maps windows-1252.xml
      through windows-1258.xml using data from
      ftp://ftp.unicode.org/Public/MAPPINGS.

2.03 2008-07-08

    - Improved POD formatting of compile_encoding and make_encmap manual pages.
      [Daniel Leidert <daniel.leidert@wgdd.de>]

2.02 2008-06-29

    - Added manual pages for compile_encoding and make_encmap.  [Daniel Leidert
      <daniel.leidert@wgdd.de>]

2.01 2007-07-24

    - Fixed crash at end of test script when using Perl 5.6.x.  (The crash was
      actually due to a bug in perl that is fixed as of Perl 5.8.0 at least, so
      the test script is simply modified to not tickle that bug.)

    - Clarified the copyright notices.  Only the *changes* in version 2.x are my
      copyright.

2.00 2007-07-18

    - Fixed bin/compile_encoding to run with Perl 5.9.x by using restricted
      hashes instead of pseudo-hashes (which have been removed from Perl 5.9.x).

    - Fixed accidental variable interpolation in Encoding.pm.

    - Fixed case of $ExtUtils::MakeMaker::VERSION in Makefile.PL so that
      ABSTRACT_FROM and AUTHOR get set when intended.

    - Added LICENSE to Makefile.PL.

1.01 1998-12-26

    - Added PREREQ_PM to Makefile.PL.

    - Added x-euc-jp-jisx0221 and x-euc-jp-unicode to maps.  [Yoshida Masato
      <yoshidam@inse.co.jp>]

    - Removed the Shift_JIS encoding and replaced it with four variations and an
      explanatory message.  [MURATA Makoto <murata@apsdc.ksp.fujixerox.co.jp>]

1.00 1998-12-13 13:43:31-05:00

    - Original version; created by h2xs 1.18.


test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..7\n"; }
END {print "not ok 1\n" unless $loaded;}
use XML::Encoding;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

# Insert your test code below (better if it prints "ok 13"
# (correspondingly "not ok 13") depending on the success of chunk 13
# of the test code):

my @prefixes = ();
my $pops = 0;
my @rnginfo = ();

sub pushpfx {
  my ($byte) = @_;

  push(@prefixes, $byte);
  undef;
}

sub poppfx {
  $pops++;
  undef;
}

sub range {
  my ($byte, $uni, $len) = @_;

  push(@rnginfo, @_);
  undef;
}

$doc1 =<<'End_of_doc;';
<encmap name="foo" expat="yes">
  <range byte='xa0' uni='x3000' len='6'/>
  <prefix byte='x81'>
    <ch byte='x41' uni='x0753'/>
    <range byte='x50' uni='x0400' len='32'/>
  </prefix>
</encmap>
End_of_doc;


my @exprng = (0xa0, 0x3000, 6, 0x41, 0x0753, 1, 0x50, 0x0400, 32);

$p = new XML::Encoding(PushPrefixFcn => \&pushpfx,
		       PopPrefixFcn  => \&poppfx,
		       RangeSetFcn   => \&range);


my $name = $p->parse($doc1);

if ($name ne 'foo') {
print "not ";
}
print "ok 2\n";

if ($prefixes[0] != 0x81) {
  print "not ";
}
print "ok 3\n";

if ($pops != @prefixes) {
  print "not ";
}
print "ok 4\n";

if (@rnginfo > @exprng) {
  print "not ";
}
print "ok 5\n";

foreach (0 .. $#exprng) {
  if ($rnginfo[$_] != $exprng[$_]) {
    print "not ";
    last;
  }
}
print "ok 6\n";

$doc1 =~ s/='32'/='200'/;

# Don't use an eval {} here to trap the parse() error
# because it causes a crash under perl-5.6.x
{
local $SIG{__DIE__} = sub {
  my $err = $_[0];
  unless ($err and $err =~ /^Len plus byte > 256/) {
    print "not ";
  }
  print "ok 7\n";
  exit;
};
$p->parse($doc1);
}

print "not ok 7\n";


encoding.h
/*****************************************************************
** encoding.h
**
** Copyright 1998 Clark Cooper
** All rights reserved.
**
** This program is free software; you can redistribute it and/or
** modify it under the same terms as Perl itself.
*/

#ifndef ENCODING_H
#define ENCODING_H 1

#define ENCMAP_MAGIC    0xfeebface

typedef struct prefixmap {
  unsigned char min;
  unsigned char len;                    /* 0 => 256 */
  unsigned short bmap_start;
  unsigned char ispfx[32];
  unsigned char ischar[32];
} PrefixMap;

typedef struct encinf
{
  unsigned short        prefixes_size;
  unsigned short        bytemap_size;
  int                   firstmap[256];
  PrefixMap             *prefixes;
  unsigned short        *bytemap;
} Encinfo;

typedef struct encmaphdr
{
  unsigned int          magic;
  char                  name[40];
  unsigned short        pfsize;
  unsigned short        bmsize;
  int                   map[256];
} Encmap_Header;

/*================================================================
** Structure of Encoding map binary encoding
**
** Note that all shorts and ints are in network order,
** so when packing or unpacking with perl, use 'n' and 'N' respectively.
** In C, use the htonl family of functions.
**
** The basic structure is:
**
**      _______________________
**      |Header (including map expat needs for 1st byte)
**      |PrefixMap * pfsize
**      |   This section isn't included for single-byte encodings.
**      |   For multiple byte encodings, when a byte represents a prefix
**      |   then it indexes into this vector instead of mapping to a
**      |   Unicode character. The PrefixMap type is declared above. The
**      |   ispfx and ischar fields are bitvectors indicating whether
**      |   the byte being mapped is a prefix or character respectively.
**      |   If neither is set, then the character is not mapped to Unicode.
**      |
**      |   The min field is the 1st byte mapped for this prefix; the
**      |   len field is the number of bytes mapped; and bmap_start is
**      |   the starting index of the map for this prefix in the overall
**      |   map (next section).
**      |unsigned short * bmsize
**      |   This section also is omitted for single-byte encodings.
**      |   Each short is either a Unicode scalar or an index into the
**      |   PrefixMap vector.
**
** The header for these files is declared above as the Encmap_Header type.
** The magic field is a magic number which should match the ENCMAP_MAGIC
** macro above. The next 40 bytes stores IANA registered name for the
** encoding. The pfsize field holds the number of PrefixMaps, which should
** be zero for single byte encodings. The bmsize field holds the number of
** shorts used for the overall map.
**
** The map field contains either the Unicode scalar encoded by the 1st byte
** or -n where n is the number of bytes that such a 1st byte implies (Expat
** requires that the number of bytes to encode a character is indicated by
** the 1st byte) or -1 if the byte doesn't map to any Unicode character.
**
** If the encoding is a multiple byte encoding, then there will be PrefixMap
** and character map sections. The 1st PrefixMap (index 0), covers a range
** of bytes that includes all 1st byte prefixes.
**
** Look at convert_to_unicode in Expat.xs to see how this data structure
** is used.
*/

#endif  /* ndef ENCODING_H */


lib/XML/Encoding.pm
################################################################
# XML::Encoding
#
# Copyright (C) 1998 Clark Cooper.  All rights reserved.
# Copyright (C) 2007-2010, 2014 Steve Hay.  All rights reserved.
#
# This module is free software; you can redistribute it and/or modify it under
# the same terms as Perl itself, i.e. under the terms of either the GNU General
# Public License or the Artistic License, as specified in the LICENCE file.
#
# See pod documentation at the end of the file
#

package XML::Encoding;

use 5.008001;

use XML::Parser;

use strict;
use warnings;

our @ISA = qw(XML::Parser);
our $VERSION = '2.12';

sub new {
  my $class = shift;
  my $self = $class->SUPER::new(@_);

  # Maybe require setting of PushPrefixFcn, PopPrefixFcn, and RangeSetFcn

  $self->setHandlers(Start => \&start, End => \&end, Final => \&fini);
  return $self;
}

sub start {
  my ($exp, $el, %attr) = @_;

  return if $exp->{EN_Skip};

  $exp->xpcroak("Root element must be encmap")
    if ($exp->depth == 0 and $el ne 'encmap');

  my $xpmode = $exp->{EN_ExpatMode};

  if ($el eq 'ch'
      or $el eq 'range')
    {
      my $byte = $attr{byte};
      $exp->xpcroak("Missing required byte attribute")
        unless defined($byte);

      $byte = cnvnumatt($exp, $byte, 'byte');
      $exp->xpcroak("byte attribute > 255") if $byte > 255;

      my $uni = $attr{uni};
      $exp->xpcroak("Missing required uni attribute")
        unless defined($uni);

      $uni = cnvnumatt($exp, $uni, 'uni');
      $exp->xpcroak("uni attribute > 0xFFFF") if $uni > 0xFFFF;

      my $len = 1;

      if ($el eq 'range') {
        $len = $attr{len};
        $exp->xpcroak("Missing required len attribute")
          unless defined($len);

        $len = cnvnumatt($exp, $len, 'len');
        $exp->xpcroak("Len plus byte > 256") if ($len + $byte) > 256;
      }

      check_range($exp, $byte, $len, $uni)
        if ($xpmode
            and $byte < 128
            and $byte != $uni
            and not $exp->in_element('prefix'));

      my $range_set_fcn = $exp->{RangeSetFcn};
      if (defined $range_set_fcn) {
        my $result = &$range_set_fcn($byte, $uni, $len);
        $exp->xpcroak($result)
          if ($xpmode and $result);
      }
    }
  elsif ($el eq 'prefix') {
    $exp->xpcroak("prefix nested too deep")
      if ($xpmode and $exp->within_element('prefix') >= 3);

    my $byte = $attr{byte};
    $exp->xpcroak("Missing required byte attribute")
      unless defined($byte);

    $byte = cnvnumatt($exp, $byte, 'byte');
    $exp->xpcroak("byte attribute > 255") if $byte > 255;
    my $push_pfx_fcn = $exp->{PushPrefixFcn};
    if (defined $push_pfx_fcn) {
      my $result = &$push_pfx_fcn($byte);
      $exp->xpcroak($result)
        if ($xpmode and $result);
    }
  }
  elsif ($el eq 'encmap') {
    my $name = $attr{name};

    $exp->xpcroak("Missing required name attribute")
      unless defined($name);

    $exp->{EN_Name} = $name;

    my $expat = $attr{expat};
    if (defined($expat)) {
      $exp->xpcroak("Value of expat attribute should be yes or no")
        unless ($expat eq 'yes' or $expat eq 'no');
      $exp->{EN_ExpatMode} = $expat eq 'yes';
    }
    else {
      $exp->{EN_ExpatMode} = 0;
    }
    $exp->xpcroak("Not an expat mode encmap")
      if ($exp->{ExpatRequired} and ! $exp->{EN_ExpatMode});
  }
  else {
    my $depth = $exp->depth;
    $exp->xpcroak($exp, "Root element isn't encmap")
      unless $depth;

    $exp->xpcarp("Skipping unrecognized element '$el'\n");
    $exp->{EN_Skip} = $depth;
  }

}  # End start

sub end {
  my ($exp, $el) = @_;

  if ($exp->{EN_Skip}) {
    $exp->{EN_Skip} = 0
      if $exp->{EN_Skip} == $exp->depth;
  }
  elsif ($el eq 'prefix') {
    my $xpmode = $exp->{EN_ExpatMode};

    my $pop_pfx_fcn = $exp->{PopPrefixFcn};
    if (defined $pop_pfx_fcn) {
      my $result = &$pop_pfx_fcn();
      $exp->xpcroak($result)
        if ($xpmode and $result);
    }
  }
}  # End end

sub fini {
  my ($exp) = @_;
  $exp->{EN_Name};
}  # End fini

sub check_range {
  my ($exp, $start, $len, $uni) = @_;

  # The following characters are exceptions to the expat rule that characters
  # in the ascii set (ordinal values < 128) must have the same value in the
  # source encoding: $@\^`{}~'
  # The ordinal values for these are 36,92,94,96,123,125,126,39
  # Any len >= 3 implies you have to be hitting some non-special
  # For 2 just check start == 125 ('}')
  # For 1 check individually.

  if ($len == 1) {
    return if chr($start) =~ /[\$@\\^`{}~']/;
  }
  elsif ($len == 2 and $start == 125) {
    return;
  }

  $exp->xpcroak("Sets ascii character to non-ascii value");
}

sub cnvnumatt {
  my ($exp, $str, $name) = @_;

  $exp->xpcroak("$name attribute is not a decimal or hex value")
    unless ($str =~ /^(?:(\d+)|x([0-9a-f]+))$/i);

  if (defined($1)) {
    return $str + 0;
  }
  else {
    return hex($2);
  }
}  # End cnvnumatt

1;

__END__

=head1 NAME

XML::Encoding - A perl module for parsing XML encoding maps.

=head1 SYNOPSIS

    use XML::Encoding;
    my $em_parser = new XML::Encoding(ErrorContext  => 2,
                                      ExpatRequired => 1,
                                      PushPrefixFcn => \&push_prefix,
                                      PopPrefixFcn  => \&pop_prefix,
                                      RangeSetFcn   => \&range_set);

    my $encmap_name = $em_parser->parsefile($ARGV[0]);

=head1 DESCRIPTION

This module, which is built as a subclass of XML::Parser, provides a parser
for encoding map files, which are XML files. The file maps/encmap.dtd in the
distribution describes the structure of these files. Calling a parse method
returns the name of the encoding map (obtained from the name attribute of
the root element). The contents of the map are processed through the
callback functions push_prefix, pop_prefix, and range_set.

=head2 Methods

This module provides no additional methods to those provided by XML::Parser,
but it does take the following additional options.

=over 4

=item * ExpatRequired

When this has a true value, then an error occurs unless the encmap
"expat" attribute is set to "yes". Whether or not the ExpatRequired option
is given, the parser enters expat mode if this attribute is set. In expat
mode, the parser checks if the encoding violates expat restrictions.

=item * PushPrefixFcn

The corresponding value should be a code reference to be called when
a prefix element starts. The single argument to the callback is an integer
which is the byte value of the prefix. An undef value should be returned
if successful. If in expat mode, a defined value causes an error and is
used as the message string.

=item * PopPrefixFcn

The corresponding value should be a code reference to be called when a
prefix element ends. No arguments are passed to this function. An undef
value should be returned if successful. If in expat mode, a defined value
causes an error and is used as the message string.

=item * RangeSetFcn

The corresponding value should be a code reference to be called when a
"range" or "ch" element is seen. The 3 arguments passed to this function are:
(byte, unicode_scalar, length)
The byte is the starting byte of a range or the byte being mapped by a
"ch" element. The unicode_scalar is the Unicode value that this byte (with
the current prefix) maps to. The length of the range is the last argument.
This will be 1 for the "ch" element. An undef value should be returned if
successful. If in expat mode, a defined value causes an error and is used
as the message string.

=back

=head1 EXPORTS

I<None>.

=head1 KNOWN BUGS

I<None>.

=head1 FEEDBACK

Patches, bug reports, suggestions or any other feedback is welcome.

Patches can be sent as GitHub pull requests at
L<https://github.com/steve-m-hay/XML-Encoding/pulls>.

Bug reports and suggestions can be made on the CPAN Request Tracker at
L<https://rt.cpan.org/Public/Bug/Report.html?Queue=XML-Encoding>.

Currently active requests on the CPAN Request Tracker can be viewed at
L<https://rt.cpan.org/Public/Dist/Display.html?Status=Active;Queue=XML-Encoding>.

Please test this distribution.  See CPAN Testers Reports at
L<https://www.cpantesters.org/> for details of how to get involved.

Previous test results on CPAN Testers Reports can be viewed at
L<https://www.cpantesters.org/distro/X/XML-Encoding.html>.

Please rate this distribution on CPAN Ratings at
L<https://cpanratings.perl.org/rate/?distribution=XML-Encoding>.

=head1 SEE ALSO

L<XML::Parser>.

=head1 AVAILABILITY

The latest version of this module is available from CPAN (see
L<perlmodlib/"CPAN"> for details) at

L<https://metacpan.org/release/XML-Encoding> or

L<https://www.cpan.org/authors/id/S/SH/SHAY/> or

L<https://www.cpan.org/modules/by-module/XML/>.

The latest source code is available from GitHub at
L<https://github.com/steve-m-hay/XML-Encoding>.

=head1 INSTALLATION

See the F<INSTALL> file.

=head1 AUTHOR

Clark Cooper E<lt>L<coopercc@netheaven.com|mailto:coopercc@netheaven.com>E<gt>.

Steve Hay E<lt>L<shay@cpan.org|mailto:shay@cpan.org>E<gt> is now maintaining
XML::Encoding as of version 2.00.

=head1 COPYRIGHT

Copyright (C) 1998 Clark Cooper.  All rights reserved.

Copyright (C) 2007-2010, 2014, 2020 Steve Hay.  All rights reserved.

=head1 LICENCE

This module is free software; you can redistribute it and/or modify it under the
same terms as Perl itself, i.e. under the terms of either the GNU General Public
License or the Artistic License, as specified in the F<LICENCE> file.

=head1 VERSION

Version 2.12

=head1 DATE

TODO

=head1 HISTORY

See the F<Changes> file.

=cut


Makefile.PL
#!perl
#===============================================================================
#
# Makefile.PL
#
# DESCRIPTION
#   Makefile creation script.
#
# COPYRIGHT
#   Copyright (C) 2014-2015 Steve Hay.  All rights reserved.
#
# LICENCE
#   This script is free software; you can redistribute it and/or modify it under
#   the same terms as Perl itself, i.e. under the terms of either the GNU
#   General Public License or the Artistic License, as specified in the LICENCE
#   file.
#
#===============================================================================

use 5.008001;

use strict;
use warnings;

use ExtUtils::MakeMaker 6.64;
use ExtUtils::MakeMaker qw(WriteMakefile);

#===============================================================================
# MAIN PROGRAM
#===============================================================================

MAIN: {
    WriteMakefile(
        NAME          => 'XML::Encoding',
        ABSTRACT_FROM => 'lib/XML/Encoding.pm',
        AUTHOR        => 'Clark Cooper <coopercc@netheaven.com>, Steve Hay <shay@cpan.org>',
        LICENSE       => 'perl_5',
        VERSION_FROM  => 'lib/XML/Encoding.pm',

        META_MERGE => {
            'meta-spec' => {
                version => 2
            },

            resources => {
                repository => {
                    type => 'git',
                    url  => 'https://github.com/steve-m-hay/XML-Encoding.git'
                }
            },

            optional_features => {
                changestest => {
                    description => 'Changes testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::CPAN::Changes' => '0'
                            }
                        }
                    }
                },

                critictest => {
                    description => 'Perl::Critic testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Perl::Critic' => '0'
                            }
                        }
                    }
                },

                podtest => {
                    description => 'POD testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Pod' => '1.00'
                            }
                        }
                    }
                },

                podcoveragetest => {
                    description => 'POD coverage testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Pod::Coverage' => '0.08'
                            }
                        }
                    }
                }
            }
        },

        MIN_PERL_VERSION => '5.008001',

        CONFIGURE_REQUIRES => {
            'ExtUtils::MakeMaker' => '6.64',
            'perl'                => '5.008001',
            'strict'              => '0',
            'warnings'            => '0'
        },

        TEST_REQUIRES => {
            'Test::More' => '0'
        },

        PREREQ_PM => {
            'XML::Parser' => '2.18',
            'fields'      => '0',
            'integer'     => '0',
            'strict'      => '0',
            'warnings'    => '0'
        },

        EXE_FILES => [qw(
            bin/compile_encoding
            bin/make_encmap
        )],

        dist => {
            PREOP   => 'find $(DISTVNAME) -type d -print|xargs chmod 0755 && ' .
                       'find $(DISTVNAME) -type f -print|xargs chmod 0644 && ' .
                       'chmod 0755 $(DISTVNAME)/bin/*',
            TO_UNIX => 'find $(DISTVNAME) -type f -print|xargs dos2unix'
        }
    );
}

#===============================================================================
